pkg/store/kotsstore/reporting_store.go

```diff
@@ -0,0 +1,236 @@
+	ON CONFLICT(created_at) DO UPDATE SET
+		license_id = EXCLUDED.license_id,
+		instance_id = EXCLUDED.instance_id,
+		cluster_id = EXCLUDED.cluster_id,
+		app_status = EXCLUDED.app_status,
+		is_kurl = EXCLUDED.is_kurl,
+		kurl_node_count_total = EXCLUDED.kurl_node_count_total,
+		kurl_node_count_ready = EXCLUDED.kurl_node_count_ready,
+		k8s_version = EXCLUDED.k8s_version,
+		kots_install_id = EXCLUDED.kots_install_id,
+		kurl_install_id = EXCLUDED.kurl_install_id,
+		is_gitops_enabled = EXCLUDED.is_gitops_enabled,
+		gitops_provider = EXCLUDED.gitops_provider,
+		cursor = EXCLUDED.cursor,
+		channel_id = EXCLUDED.channel_id,
+		channel_name = EXCLUDED.channel_name,
+		sequence = EXCLUDED.sequence,
+		source = EXCLUDED.source,
+		status = EXCLUDED.status,
+		preflight_state = EXCLUDED.preflight_state,
+		skip_preflights = EXCLUDED.skip_preflights,
+		repl_helm_installs = EXCLUDED.repl_helm_installs,
+		native_helm_installs = EXCLUDED.native_helm_installs`
+
+	statement := gorqlite.ParameterizedStatement{
+		Query: query,
+		Arguments: []interface{}{
+			createdAt.UnixMilli(),
+			licenseID,
+			reportingInfo.InstanceID,
+			reportingInfo.ClusterID,
+			reportingInfo.AppStatus,
+			reportingInfo.IsKurl,
+			reportingInfo.KurlNodeCountTotal,
+			reportingInfo.KurlNodeCountReady,
+			reportingInfo.K8sVersion,
+			reportingInfo.KOTSInstallID,
+			reportingInfo.KURLInstallID,
+			reportingInfo.IsGitOpsEnabled,
+			reportingInfo.GitOpsProvider,
+			reportingInfo.Downstream.Cursor,
+			reportingInfo.Downstream.ChannelID,
+			reportingInfo.Downstream.ChannelName,
+			reportingInfo.Downstream.Sequence,
+			reportingInfo.Downstream.Source,
+			reportingInfo.Downstream.Status,
+			reportingInfo.Downstream.PreflightState,
+			reportingInfo.Downstream.SkipPreflights,
+			reportingInfo.Downstream.ReplHelmInstalls,
+			reportingInfo.Downstream.NativeHelmInstalls,
+		},
+	}
+
+	wr, err := db.WriteOneParameterized(statement)
+	if err != nil {
+		return fmt.Errorf("failed to write instance report: %v: %v", err, wr.Err)
+	}
+
+	go func() {
+		err := s.removeOldReportingData("instance_report")
+		if err != nil {
+			logger.Warnf("failed to delete old data from instance_report: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+func (s *KOTSStore) removeOldReportingData(reportingTable string) error {
+	db := persistence.MustGetDBSession()
+
+	query := fmt.Sprintf(`select count(1) from %s`, reportingTable)
+	rows, err := db.QueryOneParameterized(gorqlite.ParameterizedStatement{
+		Query: query,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to query number of rows: %v: %v", err, rows.Err)
+	}
+
+	if !rows.Next() {
+		return ErrNotFound
+	}
+
+	var numRows int64
+	if err := rows.Scan(&numRows); err != nil {
+		return errors.Wrap(err, "failed to scan number of rows")
+	}
+
+	reportingMaxRows := int64(4000) // at 10 records per day, this is more than a year of data
+	if numRows < reportingMaxRows {
```

Comment:
should this have an `=`? otherwise, the query below might return nothing if there are exactly 4000 rows, right?
```suggestion
	if numRows <= reportingMaxRows {
```